godot-core/src/builtin/vectors/vector_macros.rs

```diff
@@ -348,139 +372,384 @@ macro_rules! impl_float_vector_component_fns {
         ($($comp:ident),*)
     ) => {
         impl $Vector {
-            pub fn lerp(self, other: Self, weight: $Scalar) -> Self {
+            /// Returns a vector with the given components.
+            pub const fn new($($comp: $Scalar),*) -> Self {
+                Self {
+                    $(
+                        $comp
+                    ),*
+                }
+            }
+
+            /// Returns a new vector with all components set to `v`.
+            pub const fn splat(v: $Scalar) -> Self {
+                Self {
+                    $(
+                        $comp: v
+                    ),*
+                }
+            }
+
+            /// Returns a new vector with all components in absolute values (i.e. positive or
+            /// zero).
+            #[inline]
+            pub fn abs(self) -> Self {
+                Self::from_glam(self.to_glam().abs())
+            }
+
+            /// Returns a new vector with all components in absolute values (i.e. positive or
+            /// zero).
+            #[inline]
+            pub fn new_abs(self) -> Self {
                 Self::new(
                     $(
-                        self.$comp.lerp(other.$comp, weight)
+                        self.$comp.abs()
                     ),*
                 )
             }
 
-            pub fn bezier_derivative(self, control_1: Self, control_2: Self, end: Self, t: $Scalar) -> Self {
-                $(
-                    let $comp = self.$comp.bezier_derivative(control_1.$comp, control_2.$comp, end.$comp, t);
-                )*
+            /// Returns a new vector with all components clamped between the components of `min` and `max`.
+            ///
+            /// Panics
+            /// Panics if `min` > `max`, `min` is NaN, or `max` is NaN.
+            #[inline]
+            pub fn clamped(self, min: Self, max: Self) -> Self {
+                Self::new(
+                    $(
+                        self.$comp.clamp(min.$comp, max.$comp)
+                    ),*
+                )
+            }
 
-                Self::new($($comp),*)
+            /// Returns the dot product of this vector and `with`.
+            #[inline]
+            pub fn dot(self, with: Self) -> $Scalar {
+                self.to_glam().dot(with.to_glam())
             }
 
-            pub fn bezier_interpolate(self, control_1: Self, control_2: Self, end: Self, t: $Scalar) -> Self {
-                $(
-                    let $comp = self.$comp.bezier_interpolate(control_1.$comp, control_2.$comp, end.$comp, t);
-                )*
+            /// Returns the length (magnitude) of this vector.
+            #[inline]
+            pub fn new_length(self) -> real {
+                (strip_leading_plus!($( + (self.$comp * self.$comp) )*) as real).sqrt()
```

Comment:
I guess `$( (self.$comp * self.$comp) )+*` doesn't work?

Also, `as real` should be replaced with [`RealConv`](https://godot-rust.github.io/docs/gdext/master/godot/builtin/trait.RealConv.html) functions.

Reply:
Yes, the `+` gets interpretet as [any number of repetitions but at least one](https://doc.rust-lang.org/reference/macros-by-example.html#repetitions) and you can't escape it.

I don't quite understand how I should be replacing `as real` with `RealConv`, since the return type is also `real`.

Reply:
The `impl_vector_fns` macro implements functions for floating-point and integer vectors. Because the result of `strip_leading_plus!($( + (self.$comp * self.$comp) )*)` for integer vectors is `i32` and `sqrt` doesn't exists on `i32` it needs to be casted. Since there is no performance impact for floating-point vectors i didn't see the need to split this, as the only difference would be the type casting.